File: 07_trees/trees.rs
//! TREES  ·  Rust
//! BST + common tree problems using Box<TreeNode>.
//! Rust trees use Option<Box<TreeNode>> as the nullable owned pointer type.

use std::collections::VecDeque;

#[derive(Debug, Clone, PartialEq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Box<TreeNode>>,
    pub right: Option<Box<TreeNode>>,
}
impl TreeNode {
    pub fn new(val: i32) -> Box<Self> { Box::new(TreeNode { val, left: None, right: None }) }
}
pub type Tree = Option<Box<TreeNode>>;

/// Build from level-order slice (0 = null sentinel here we use Option in caller)
pub fn from_vec(vals: &[Option<i32>]) -> Tree {
    if vals.is_empty() || vals[0].is_none() { return None; }
    let root = Box::new(TreeNode { val: vals[0].unwrap(), left: None, right: None });
    let mut queue: VecDeque<*mut TreeNode> = VecDeque::new();
    queue.push_back(Box::into_raw(root.clone()) as *mut TreeNode);
    // Rebuild properly using an arena-free approach
    fn build(vals: &[Option<i32>], i: usize) -> Tree {
        if i >= vals.len() || vals[i].is_none() { return None; }
        Some(Box::new(TreeNode {
            val: vals[i].unwrap(),
            left: build(vals, 2*i+1),
            right: build(vals, 2*i+2),
        }))
    }
    build(vals, 0)
}

pub fn inorder(root: &Tree) -> Vec<i32> {
    let mut res = vec![];
    fn dfs(n: &Tree, r: &mut Vec<i32>) {
        if let Some(node) = n { dfs(&node.left,r); r.push(node.val); dfs(&node.right,r); }
    }
    dfs(root, &mut res); res
}

pub fn level_order(root: &Tree) -> Vec<Vec<i32>> {
    let mut res = vec![];
    if root.is_none() { return res; }
    let mut q: VecDeque<&Box<TreeNode>> = VecDeque::new();
    q.push_back(root.as_ref().unwrap());
    while !q.is_empty() {
        let len = q.len();
        let mut level = vec![];
        for _ in 0..len {
            let n = q.pop_front().unwrap();
            level.push(n.val);
            if let Some(ref l) = n.left { q.push_back(l); }
            if let Some(ref r) = n.right { q.push_back(r); }
        }
        res.push(level);
    }
    res
}

pub fn max_depth(root: &Tree) -> i32 {
    match root { None => 0, Some(n) => 1 + max_depth(&n.left).max(max_depth(&n.right)) }
}

pub fn is_symmetric(root: &Tree) -> bool {
    fn mirror(l: &Tree, r: &Tree) -> bool {
        match (l, r) {
            (None, None) => true,
            (Some(a), Some(b)) => a.val==b.val && mirror(&a.left,&b.right) && mirror(&a.right,&b.left),
            _ => false,
        }
    }
    match root { None => true, Some(n) => mirror(&n.left, &n.right) }
}

pub fn max_path_sum(root: &Tree) -> i32 {
    let mut best = i32::MIN;
    fn gain(n: &Tree, best: &mut i32) -> i32 {
        if let Some(node) = n {
            let l = gain(&node.left, best).max(0);
            let r = gain(&node.right, best).max(0);
            *best = (*best).max(l + r + node.val);
            node.val + l.max(r)
        } else { 0 }
    }
    gain(root, &mut best); best
}

pub fn is_valid_bst(root: &Tree) -> bool {
    fn v(n: &Tree, lo: i64, hi: i64) -> bool {
        match n {
            None => true,
            Some(node) => {
                let val = node.val as i64;
                val > lo && val < hi && v(&node.left, lo, val) && v(&node.right, val, hi)
            }
        }
    }
    v(root, i64::MIN, i64::MAX)
}

pub fn diameter(root: &Tree) -> i32 {
    let mut best = 0;
    fn h(n: &Tree, best: &mut i32) -> i32 {
        if let Some(node) = n {
            let l = h(&node.left, best); let r = h(&node.right, best);
            *best = (*best).max(l + r); 1 + l.max(r)
        } else { 0 }
    }
    h(root, &mut best); best
}

pub fn right_side_view(root: &Tree) -> Vec<i32> {
    let mut res = vec![];
    if root.is_none() { return res; }
    let mut q: VecDeque<&Box<TreeNode>> = VecDeque::new();
    q.push_back(root.as_ref().unwrap());
    while !q.is_empty() {
        let len = q.len();
        for i in 0..len {
            let n = q.pop_front().unwrap();
            if i == len-1 { res.push(n.val); }
            if let Some(ref l) = n.left { q.push_back(l); }
            if let Some(ref r) = n.right { q.push_back(r); }
        }
    }
    res
}

pub fn is_balanced(root: &Tree) -> bool {
    fn h(n: &Tree) -> i32 {
        if let Some(node) = n {
            let l = h(&node.left); let r = h(&node.right);
            if l < 0 || r < 0 || (l-r).abs() > 1 { return -1; }
            1 + l.max(r)
        } else { 0 }
    }
    h(root) >= 0
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_inorder() {
        let t = from_vec(&[Some(2),Some(1),Some(3)]);
        assert_eq!(inorder(&t), vec![1,2,3]);
    }
    #[test]
    fn test_max_depth() {
        assert_eq!(max_depth(&from_vec(&[Some(3),Some(9),Some(20),None,None,Some(15),Some(7)])), 3);
        assert_eq!(max_depth(&None), 0);
    }
    #[test]
    fn test_is_symmetric() {
        assert!(is_symmetric(&from_vec(&[Some(1),Some(2),Some(2),Some(3),Some(4),Some(4),Some(3)])));
        assert!(!is_symmetric(&from_vec(&[Some(1),Some(2),Some(2),None,Some(3),None,Some(3)])));
    }
    #[test]
    fn test_max_path_sum() {
        assert_eq!(max_path_sum(&from_vec(&[Some(-10),Some(9),Some(20),None,None,Some(15),Some(7)])), 42);
    }
    #[test]
    fn test_is_valid_bst() {
        assert!(is_valid_bst(&from_vec(&[Some(2),Some(1),Some(3)])));
        assert!(!is_valid_bst(&from_vec(&[Some(5),Some(1),Some(4),None,None,Some(3),Some(6)])));
    }
    #[test]
    fn test_diameter() {
        assert_eq!(diameter(&from_vec(&[Some(1),Some(2),Some(3),Some(4),Some(5)])), 3);
    }
    #[test]
    fn test_right_side_view() {
        assert_eq!(right_side_view(&from_vec(&[Some(1),Some(2),Some(3),None,Some(5),None,Some(4)])), vec![1,3,4]);
    }
    #[test]
    fn test_is_balanced() {
        assert!(is_balanced(&from_vec(&[Some(3),Some(9),Some(20),None,None,Some(15),Some(7)])));
        assert!(!is_balanced(&from_vec(&[Some(1),Some(2),Some(2),Some(3),Some(3),None,None,Some(4),Some(4)])));
    }
}

File: 07_trees/trees.js
"use strict";

/**
 * ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 * TREES  ·  JavaScript
 * ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 * BST: left < node < right. O(log n) avg, O(n) worst.
 * Traversals: inorder(sorted), preorder, postorder, BFS level-order.
 */

// ┌─────────────────────────────────────────────────────────────────┐
// │ TABLE OF CONTENTS                                               │
// ├─────────────────────────────────────────────────────────────────┤
// │ 1. TreeNode class                                               │
// │ 2. fromArray helper                                             │
// │ 3. BST class                                                    │
// │ 4. Traversals                                                   │
// │    - inorder, levelOrder                                        │
// │ 5. Problems                                                     │
// │    - maxDepth, invertTree, isSymmetric, isValidBST      🟢      │
// │    - lowestCommonAncestor, rightSideView, kthSmallest   🟡      │
// │    - buildTreeFromPreIn, diameterOfBinaryTree           🟡      │
// │    - isBalanced, zigzagLevelOrder                      🟡      │
// │    - maxPathSum                                         🔴      │
// │ 6. Tests                                                       │
// └─────────────────────────────────────────────────────────────────┘

class TreeNode {
  /**
   * @param {number} val
   */
  constructor(val) {
    this.val = val;
    /** @type {TreeNode|null} */
    this.left = null;
    /** @type {TreeNode|null} */
    this.right = null;
  }
}

// ── Build helper ───────────────────────────────

/** @param {(number|null)[]} vals
 * @returns {TreeNode|null}
 */
function fromArray(vals) {
  if (!vals.length || vals[0] == null) return null;
  const root = new TreeNode(vals[0]);
  const q = [root];
  let i = 1;
  while (q.length && i < vals.length) {
    const node = q.shift();
    if (i < vals.length && vals[i] != null) {
      node.left = new TreeNode(vals[i]);
      q.push(node.left);
    }
    i++;
    if (i < vals.length && vals[i] != null) {
      node.right = new TreeNode(vals[i]);
      q.push(node.right);
    }
    i++;
  }
  return root;
}

// ── BST ───────────────────────────────────────

class BST {
  constructor() {
    /** @type {TreeNode|null} */
    this.root = null;
  }

  /** @param {number} v */
  insert(v) {
    this.root = this._ins(this.root, v);
  }

  /** @param {TreeNode|null} n
   * @param {number} v
   * @returns {TreeNode}
   */
  _ins(n, v) {
    if (!n) return new TreeNode(v);
    if (v < n.val) n.left = this._ins(n.left, v);
    else if (v > n.val) n.right = this._ins(n.right, v);
    return n;
  }

  /** @returns {number[]} */
  inorder() {
    const r = [];
    const dfs = (n) => {
      if (!n) return;
      dfs(n.left);
      r.push(n.val);
      dfs(n.right);
    };
    dfs(this.root);
    return r;
  }
}

// ── Traversals ────────────────────────────────

/** @param {TreeNode|null} root
 * @returns {number[]}
 */
function inorder(root) {
  const r = [];
  const dfs = (n) => {
    if (!n) return;
    dfs(n.left);
    r.push(n.val);
    dfs(n.right);
  };
  dfs(root);
  return r;
}

/** @param {TreeNode|null} root
 * @returns {number[][]}
 */
function levelOrder(root) {
  if (!root) return [];
  const res = [],
    q = [root];
  while (q.length) {
    const level = [],
      len = q.length;
    for (let i = 0; i < len; i++) {
      const n = q.shift();
      level.push(n.val);
      if (n.left) q.push(n.left);
      if (n.right) q.push(n.right);
    }
    res.push(level);
  }
  return res;
}

// ══════════════════════════════════════
// PROBLEMS
// ══════════════════════════════════════

/** 🟢 Maximum Depth of Binary Tree (LC #104)
 * @param {TreeNode|null} r
 * @returns {number}
 */
function maxDepth(r) {
  return r ? 1 + Math.max(maxDepth(r.left), maxDepth(r.right)) : 0;
}

/** 🟢 Invert Binary Tree (LC #226)
 * @param {TreeNode|null} r
 * @returns {TreeNode|null}
 */
function invertTree(r) {
  if (!r) return null;
  [r.left, r.right] = [invertTree(r.right), invertTree(r.left)];
  return r;
}

/** 🟢 Symmetric Tree (LC #101)
 * @param {TreeNode|null} r
 * @returns {boolean}
 */
function isSymmetric(r) {
  const m = (l, ri) =>
    !l && !ri ? true : !l || !ri ? false : l.val === ri.val && m(l.left, ri.right) && m(l.right, ri.left);
  return r ? m(r.left, r.right) : true;
}

/** 🟡 Lowest Common Ancestor (LC #236)
 * @param {TreeNode|null} r
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @returns {TreeNode|null}
 */
function lowestCommonAncestor(r, p, q) {
  if (!r || r === p || r === q) return r;
  const l = lowestCommonAncestor(r.left, p, q),
    ri = lowestCommonAncestor(r.right, p, q);
  return l && ri ? r : l ?? ri;
}

/** 🔴 Binary Tree Maximum Path Sum (LC #124)
 * @param {TreeNode|null} r
 * @returns {number}
 */
function maxPathSum(r) {
  let best = -Infinity;
  const g = (n) => {
    if (!n) return 0;
    const l = Math.max(g(n.left), 0),
      ri = Math.max(g(n.right), 0);
    best = Math.max(best, l + ri + n.val);
    return n.val + Math.max(l, ri);
  };
  g(r);
  return best;
}

/** 🟡 Binary Tree Right Side View (LC #199)
 * @param {TreeNode|null} r
 * @returns {number[]}
 */
function rightSideView(r) {
  const res = [];
  if (!r) return res;
  const q = [r];
  while (q.length) {
    const len = q.length;
    for (let i = 0; i < len; i++) {
      const n = q.shift();
      if (i === len - 1) res.push(n.val);
      if (n.left) q.push(n.left);
      if (n.right) q.push(n.right);
    }
  }
  return res;
}

/** 🟡 Kth Smallest Element in BST (LC #230)
 * @param {TreeNode|null} r
 * @param {number} k
 * @returns {number}
 */
function kthSmallest(r, k) {
  const stack = [];
  let cur = r,
    cnt = 0;
  while (stack.length || cur) {
    while (cur) {
      stack.push(cur);
      cur = cur.left;
    }
    cur = stack.pop();
    if (++cnt === k) return cur.val;
    cur = cur.right;
  }
  return -1;
}

/** 🟢 Validate Binary Search Tree (LC #98)
 * @param {TreeNode|null} r
 * @returns {boolean}
 */
function isValidBST(r) {
  const v = (n, lo, hi) =>
    !n ? true : n.val > lo && n.val < hi && v(n.left, lo, n.val) && v(n.right, n.val, hi);
  return v(r, -Infinity, Infinity);
}

/** 🟡 Construct Binary Tree from Preorder and Inorder (LC #105)
 * @param {number[]} pre
 * @param {number[]} ino
 * @returns {TreeNode|null}
 */
function buildTreeFromPreIn(pre, ino) {
  if (!pre.length) return null;
  const root = new TreeNode(pre[0]),
    mid = ino.indexOf(pre[0]);
  root.left = buildTreeFromPreIn(pre.slice(1, mid + 1), ino.slice(0, mid));
  root.right = buildTreeFromPreIn(pre.slice(mid + 1), ino.slice(mid + 1));
  return root;
}

/** 🟢 Diameter of Binary Tree (LC #543)
 * @param {TreeNode|null} r
 * @returns {number}
 */
function diameterOfBinaryTree(r) {
  let best = 0;
  const h = (n) => {
    if (!n) return 0;
    const l = h(n.left),
      ri = h(n.right);
    best = Math.max(best, l + ri);
    return 1 + Math.max(l, ri);
  };
  h(r);
  return best;
}

/** 🟡 Balanced Binary Tree (LC #110)
 * @param {TreeNode|null} r
 * @returns {boolean}
 */
function isBalanced(r) {
  const h = (n) => {
    if (!n) return 0;
    const l = h(n.left),
      ri = h(n.right);
    if (l < 0 || ri < 0 || Math.abs(l - ri) > 1) return -1;
    return 1 + Math.max(l, ri);
  };
  return h(r) >= 0;
}

/** 🟡 Binary Tree Zigzag Level Order (LC #103)
 * @param {TreeNode|null} r
 * @returns {number[][]}
 */
function zigzagLevelOrder(r) {
  const res = [];
  if (!r) return res;
  const q = [r];
  let leftToRight = true;
  while (q.length) {
    const len = q.length;
    const level = [];
    for (let i = 0; i < len; i++) {
      const n = q.shift();
      leftToRight ? level.push(n.val) : level.unshift(n.val);
      if (n.left) q.push(n.left);
      if (n.right) q.push(n.right);
    }
    res.push(level);
    leftToRight = !leftToRight;
  }
  return res;
}

// ══════════════════════════════════════
// TESTS
// ══════════════════════════════════════

/** @param {boolean} c
 * @param {string} m
 */
function assert(c, m) {
  if (!c) throw new Error(`FAIL: ${m}`);
}

/** @param {unknown} a
 * @param {unknown} b
 * @returns {boolean}
 */
const eq = (a, b) => JSON.stringify(a) === JSON.stringify(b);

function runTests() {
  console.log("Running tree tests...\n");

  const bst = new BST();
  [5, 3, 7, 1, 4, 6, 8].forEach((v) => bst.insert(v));
  assert(eq(bst.inorder(), [1, 3, 4, 5, 6, 7, 8]), "bst inorder");
  console.log("  ✅ BST: insert / inorder");

  const t = fromArray([3, 9, 20, null, null, 15, 7]);
  assert(maxDepth(t) === 3, "maxDepth");
  assert(eq(levelOrder(t), [[3], [9, 20], [15, 7]]), "levelOrder");
  console.log("  ✅ maxDepth / levelOrder");

  const inv = fromArray([4, 2, 7, 1, 3, 6, 9]);
  invertTree(inv);
  assert(inorder(inv)[0] === 9, "invertTree");
  console.log("  ✅ invertTree");

  assert(isSymmetric(fromArray([1, 2, 2, 3, 4, 4, 3])), "symmetric");
  assert(!isSymmetric(fromArray([1, 2, 2, null, 3, null, 3])), "asymmetric");
  console.log("  ✅ isSymmetric");

  assert(isValidBST(fromArray([2, 1, 3])), "validBST");
  assert(!isValidBST(fromArray([5, 1, 4, null, null, 3, 6])), "invalidBST");
  console.log("  ✅ isValidBST");

  assert(maxPathSum(fromArray([-10, 9, 20, null, null, 15, 7])) === 42, "maxPathSum");
  console.log("  ✅ maxPathSum");

  assert(eq(rightSideView(fromArray([1, 2, 3, null, 5, null, 4])), [1, 3, 4]), "rightSideView");
  console.log("  ✅ rightSideView");

  const bst3 = new BST();
  [3, 1, 4, null, 2].forEach((v) => v != null && bst3.insert(v));
  assert(kthSmallest(bst3.root, 1) === 1, "kthSmallest");
  console.log("  ✅ kthSmallest");

  assert(diameterOfBinaryTree(fromArray([1, 2, 3, 4, 5])) === 3, "diameter");
  console.log("  ✅ diameterOfBinaryTree");

  assert(isBalanced(fromArray([3, 9, 20, null, null, 15, 7])), "balanced");
  assert(!isBalanced(fromArray([1, 2, 2, 3, 3, null, null, 4, 4])), "unbalanced");
  console.log("  ✅ isBalanced");

  assert(
    eq(
      buildTreeFromPreIn([3, 9, 20, 15, 7], [9, 3, 15, 20, 7]),
      fromArray([3, 9, 20, null, null, 15, 7])
    ),
    "buildFromPreIn"
  );
  console.log("  ✅ buildFromPreIn");

  assert(
    eq(zigzagLevelOrder(fromArray([3, 9, 20, null, null, 15, 7])), [
      [3],
      [20, 9],
      [15, 7],
    ]),
    "zigzag"
  );
  console.log("  ✅ zigzagLevelOrder");

  console.log("\n✓ trees — all tests passed");
}
runTests();
